src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelTrace.cs

```diff
@@ -117,6 +117,10 @@ internal class KestrelTrace : IKestrelTrace
             LoggerMessage.Define<string>(LogLevel.Debug, new EventId(40, nameof(Http2MaxConcurrentStreamsReached)),
                 @"Connection id ""{ConnectionId}"" reached the maximum number of concurrent HTTP/2 streams allowed.");
 
+        private static readonly Action<ILogger, Exception> _invalidResponseHeaderRemoved =
+            LoggerMessage.Define(LogLevel.Debug, new EventId(41, nameof(InvalidResponseHeaderRemoved)),
+                "A response header has been removed because it was invalid for the current protocol.");
```

Comment:
Can this log message be a little more specific? Maybe:

```suggestion
                "One or more of the following response headers have been removed because it they are invalid for HTTP/2 and HTTP/3 responses: 'Connection', 'Transfer-Encoding', 'Keep-Alive', 'Upgrade' and 'Proxy-Connection'");
```

Reply:
And can we make this a warning?

Reply:
Is it really important enough for a warning? We've mitigated the issue, the app can largely ignore it. If you did have a component blindly setting one of these headers you'd end up getting warning on every request.

Reply:
If this was a client issue the app had no control over a debug log makes sense, but this is the application trying to do something invalid. I'm not sure we should assume ignoring application-set response headers is completely benign even if it usually is. What if the developer expected setting one of these headers to actually do something? In my experience, people don't turn on debug logs until they've exhausted all other options. Most of the time, people file a GitHub issues and/or contact support first.

Reply:
> In my experience, people don't turn on debug logs until they've exhausted all other options. Most of the time, people file a GitHub issues and/or contact support first.

True, but they also file issues when we write noisy logs. I don't know where the balance is. I erred on a lower level because you can't assume a developer controls every aspect of their app and would be able to fix these. These headers could be coming from 3rd party libraries/middleware they're consuming.